thewalrus/reference.py

```diff
@@ -278,3 +282,121 @@ def hafnian(M, loop=False):
         tot_sum = tot_sum + result
 
     return tot_sum
+
+
+def mapper(x, objects):
+    """Helper function to turn a permutation and bistring into an element of rpmp.
+
+    Args:
+        x (tuple): tuple containing a permutation and a bistring.
+        objects (list): list objects to permute
+
+    Returns:
+        tuple: permuted objects
+    """
+    (perm, bit) = x
+    m = len(bit)
+    Blist = [list(range(m)), list(range(m, 2 * m))]
+    for i, j in enumerate(bit):
+        if int(j):
+            (Blist[0][i], Blist[1][i]) = (Blist[1][i], Blist[0][i])
+    Blist = [Blist[0][i] for i in tuple((0,) + perm)] + [Blist[1][i] for i in tuple((0,) + perm)]
+    dico_list = {j: i + 1 for i, j in enumerate(Blist)}
+    new_mapping_list = {
+        objects[dico_list[i] - 1]: objects[dico_list[j] - 1]
+        for i, j in zip(list(range(0, m - 1)) + [m], list(range(m + 1, 2 * m)) + [m - 1])
+    }
+    return tuple(new_mapping_list.items())
+
+
+def bitstrings(n):
+    """Returns the bistrings from 0 to n/2
+
+    Args:
+        n (int) : Twice the highest bitstring value.
+
+    Returns:
+        (iterator) : An iterable of all bistrings.
+    """
+    for binary in map("".join, product("01", repeat=n - 1)):
+        yield "0" + binary
+
+
+def rpmp(s):
+    """Generates the restricted set of perfect matchings matching permutations.
+
+    Args:
+        s (tuple): tuple of labels to be used
+
+    Returns:
+        generator: the set of restricted perfect matching permutations of the tuple s
+    """
+    m = len(s) // 2
+
+    def local_mapper(x):
+        """Helper function to define a local mapper based on the symbols s
+        Args:
+            x (iterable): object to be mapped
+        """
+        return mapper(x, s)
+
+    for i in product(permutations(range(1, m)), bitstrings(m)):
+        yield local_mapper(i)
+
+
+def splitter(elem):
+    """Takes an element from rpmp and returns all the associated elements in rspm
```

Comment:
Avoid using acronyms: someone without context (like me) will struggle hard to understand their meaning when reading the docstrings.